onnxruntime/core/optimizer/bias_softmax_fusion.cc

```diff
@@ -0,0 +1,269 @@
+      if (input1->Shape()->dim(N1-i) != input2->Shape()->dim(N2-i)) {
+        singlebatch_shape_matches = false;
+        break;
+      }
+    }
+
+    if (!singlebatch_shape_matches) {
+      return false;
+    }
+
+    // identify broadcast dimensions (i.e. B to k-1 in expression above)
+    // also distinguish between input and mask in this process
+    bool mask_can_simple_broadcast = true;
+
+    int B;
+
+     // case 1: mask rank == input rank
+    if (N1 == N2) {
+
+      // discover B (first axis where shapes don't match)
+      B = 0;
+      while (B < k && input1->Shape()->dim(B) == input2->Shape()->dim(B)) {
+        B++;
+      }
+
+      // use B dimension to distinguish input and mask
+      select_input_on_lhs_condition(input1->Shape()->dim(B) != 1, add_node, input, mask);
+
+      // confirm mask dimensions are ones on broadcast axes B to (k-1)
+      for (int i = B; i < k; i++) {
+        if (mask->Shape()->dim(i) != 1) {
+          mask_can_simple_broadcast = false;
+          break;
+        }
+      }
+    }
+
+    // case 2: mask rank < input rank
+    else { 
+
+      B = 0;
+      select_input_on_lhs_condition(N1 > N2, add_node, input, mask);
+
+      // confirm any mask dimensions are ones before softmax axis
+      int mask_rank = mask->Shape()->dim_size();
+      for (int i = 0; i < mask_rank - singlebatch_rank; i++) {
+        if (mask->Shape()->dim(i) != 1) {
+          mask_can_simple_broadcast = false;
+          break;
+        }
+      }
+    }
+
+    if (!mask_can_simple_broadcast) {
+      return false;
+    }
+
+    broadcast_axis = B;
+    softmax_axis = k;
+    return true;
+}
+
+// coalesce subgraph nodes into fused node
+void FuseBiasSoftmaxSubgraph(
+  Graph& graph, 
+  Node& add_node, 
+  Node& softmax_node, 
+  NodeArg *input, 
+  NodeArg *mask,
+  int broadcast_axis,
+  int softmax_axis) {
+
+    std::vector<NodeArg*> fused_inputs{input, mask};
+
+    std::string fused_desc = 
+      "fused " + add_node.Name() + " and " + softmax_node.Name() + " into softmax(input + bias)";
+
+    std::string op_type = "BiasSoftmax";
+    Node& fused_node = graph.AddNode(graph.GenerateNodeName(op_type),
+                                    op_type,
+                                    fused_desc,
+                                    fused_inputs,
+                                    {},
+                                    {},
+                                    kMSDomain);
+
+    // add softmax axis and broadcast axis (to simplify runtime logic)
+    // recall broadcast along axes B ... (k-1)
+    fused_node.AddAttribute("softmax_axis", static_cast<int64_t>(softmax_axis));
+    fused_node.AddAttribute("broadcast_axis", static_cast<int64_t>(broadcast_axis));
+
+    // finalize node fusion (e.g. remove old nodes and shift outputs)
+    fused_node.SetExecutionProviderType(add_node.GetExecutionProviderType());
+    graph_utils::FinalizeNodeFusion(graph, {add_node, softmax_node}, fused_node);
+}
```

Comment:
general comment: per https://google.github.io/styleguide/cppguide.html#Unnamed_Namespaces_and_Static_Variables, consider making local helper functions static or putting them in the unnamed namespace

Reply:
e.g., consider ```namespace { ... }```